coderd/insights.go

```diff
@@ -43,3 +51,390 @@ func (api *API) deploymentDAUs(rw http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	// Fetch all users so that we can still include users that have no
+	// latency data.
+	users, err := api.Database.GetUsers(ctx, database.GetUsersParams{})
+	if err != nil {
+		httpapi.InternalServerError(rw, err)
+		return
+	}
+
+	templateIDSet := make(map[uuid.UUID]struct{})
+	usersWithLatencyByID := make(map[uuid.UUID]codersdk.UserLatency)
+	for _, row := range rows {
+		for _, templateID := range row.TemplateIDs {
+			templateIDSet[templateID] = struct{}{}
+		}
+		usersWithLatencyByID[row.UserID] = codersdk.UserLatency{
+			TemplateIDs: row.TemplateIDs,
+			UserID:      row.UserID,
+			Username:    row.Username,
+			LatencyMS: &codersdk.ConnectionLatency{
+				P50: row.WorkspaceConnectionLatency50,
+				P95: row.WorkspaceConnectionLatency95,
+			},
+		}
+	}
+	userLatencies := []codersdk.UserLatency{}
+	for _, user := range users {
+		userLatency, ok := usersWithLatencyByID[user.ID]
+		if !ok {
+			// We only include deleted/inactive users if they were
+			// active as part of the requested timeframe.
+			if user.Deleted || user.Status != database.UserStatusActive {
+				continue
+			}
+
+			userLatency = codersdk.UserLatency{
+				TemplateIDs: []uuid.UUID{},
+				UserID:      user.ID,
+				Username:    user.Username,
+			}
+		}
+		userLatencies = append(userLatencies, userLatency)
+	}
+
+	// TemplateIDs that contributed to the data.
+	seenTemplateIDs := make([]uuid.UUID, 0, len(templateIDSet))
+	for templateID := range templateIDSet {
+		seenTemplateIDs = append(seenTemplateIDs, templateID)
+	}
+	slices.SortFunc(seenTemplateIDs, func(a, b uuid.UUID) bool {
+		return a.String() < b.String()
+	})
+
+	resp := codersdk.UserLatencyInsightsResponse{
+		Report: codersdk.UserLatencyInsightsReport{
+			StartTime:   startTime,
+			EndTime:     endTime,
+			TemplateIDs: seenTemplateIDs,
+			Users:       userLatencies,
+		},
+	}
+	httpapi.Write(ctx, rw, http.StatusOK, resp)
+}
+
+// @Summary Get insights about templates
+// @ID get-insights-about-templates
+// @Security CoderSessionToken
+// @Produce json
+// @Tags Insights
+// @Success 200 {object} codersdk.TemplateInsightsResponse
+// @Router /insights/templates [get]
+func (api *API) insightsTemplates(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	if !api.Authorize(r, rbac.ActionRead, rbac.ResourceDeploymentValues) {
+		httpapi.Forbidden(rw)
+		return
+	}
+
+	p := httpapi.NewQueryParamParser().
+		Required("start_time").
+		Required("end_time")
+	vals := r.URL.Query()
+	var (
+		// The QueryParamParser does not preserve timezone, so we need
+		// to parse the time ourselves.
+		startTimeString = p.String(vals, "", "start_time")
+		endTimeString   = p.String(vals, "", "end_time")
+		intervalString  = p.String(vals, string(codersdk.InsightsReportIntervalNone), "interval")
+		templateIDs     = p.UUIDs(vals, []uuid.UUID{}, "template_ids")
+	)
+	p.ErrorExcessParams(vals)
+	if len(p.Errors) > 0 {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
```

Comment:
I can add a few tests for the bad requests, but I'd like to avoid creating a precedent for testing excess params here. It's not something we should test per-endpoint IMO and it's also pretty hard to test currently (we use codersdk a lot and we can't tell it to send nonsense params), we also mostly avoid internal tests which is required to test the handler directly.

Reply:
The reason why I raised this is because we don't depend on validation from a library, but there are some custom rules for timestamps. Feel free to ignore it if you don't consider it to be a high priority.

Reply:
I added a few tests for the endpoints, I can't directly test for e.g. malformed input this way, but the full chain is at least tested. I.e. we ensure that the validation function is being called and then we have internal tests for the validation function as well.